### PR TITLE
Support setting Multus default network via annotations.

### PR DESCRIPTION
**Pull request by greg-bock**
**What this PR does / why we need it**:
Adds support for setting the default network in Multus via annotations. https://github.com/intel/multus-cni/pull/174

This adds a new `Default` field to the `CniNetwork` struct.
This can only be set once and is enforced by `validating-webhook`.
Setting this field on a Network in the VMI spec will cause the `v1.multus-cni.io/default-network` annotation to be added to the launcher pod with the network name with the `default` setting.
All other changes are to support naming the default network `eth` vs the normal multus prefix of `net` and fixing count logic for numeric suffix of the network/device name. 

This is useful if you want the vmi default network to not be the default network in a Multus networking setup:
- Users might want to have a separate pod network from the default network to segregate workload networking from k8s or other sensitive services.
- Users might want to only have a single network from Multus that isn't a pod network at all (External SRIOV)  

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Adds support for setting the default network in Multus via annotations.
action required: A version of the Multus cni that supports default via annotations.
```

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically.
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by greg-bock** [comment]
@booxter @SchSeba this replaces PR #1807 and is based off master and not my SR-IOV branch.

**Comment by dankenigsberg** [comment]
@greg-bock sorry for the slow response here. I must admit that I'm not a huge fan of intel/multus-cni#174, as it makes it too easy to break the rule that all kubernetes Pods sit on the same flat pingable network.

Regardless, I did not understand why this is useful for kubevirt

>     * Users might want to only have a single network from Multus that isn't a pod network at all (External SRIOV)

in that case, I'd just leave the default interface on the Pod (not refer to `default` in the VM spec). In what way is your approach better? qemu should give us enough protection to avoid the guest code reaching to the default network.

**Comment by greg-bock** [comment]
We have a few different use cases and concerns, but the major ones are defense in depth and multi tenancy.

**Comment by dankenigsberg** [comment]
> defense in depth and multi tenancy.

It may be non-native English, but I don't understand your answer. I'd appreciate an elaboration.

I, however, thought of an answer of my own: my personal liking of intel/multus-cni#174 does not matter much. It is fact that this is now possible with Pods, and as such we kinda have to make it possible with VMs. @booxter, what do you say of this pr?

**Comment by greg-bock** [comment]
> It may be non-native English, but I don't understand your answer. I'd appreciate an elaboration.

VM escapes from qemu aren't unheard of, if something did happen to break out of qemu into the kubevirt launcher pod it would now have access to a network it previously did not. We would like to avoid that possibility.

**Comment by fabiand** [comment]
> VM escapes from qemu aren't unheard of, if something did happen to break out of qemu into the kubevirt launcher pod it would now have access to a network it previously did not. 

Usually all networks the VM pod is connected to, can also be reached from the VM.
Thus, even if the user can escape the VM, or if the user plainly access the VM, he should not have access to more networks than from within the VM.
Thus could you explain a little why you are seeing that a user could gain access to more networks from the VM pod than from the VM itself?

One design principle we are following (and working on to achieve) is that a user should never have more privileges/access from the VM *pod*, than he has from within the VM.

> We would like to avoid that possibility.

**Comment by dankenigsberg** [comment]
> One design principle we are following (and working on to achieve) is that a user should never have more privileges/access from the VM pod, than he has from within the VM.

As far as I can see, that's exactly what this PR is enforcing. There is a user that does not need access to the default network from her VM; this patch makes sure that she does not have this access from her VM Pod.

**Comment by booxter** [comment]
@greg-bock could you please clarify why AutoattachPodInterface is not enough for your needs? Is it because pod network is still present in the pod and you don't want it there? Is there a reason for this detached interface to exist inside the pod at all? Can we kill it when preparing the launcher pod networking?

**Comment by greg-bock** [comment]
> It is fact that this is now possible with Pods, and as such we kinda have to make it possible with VMs. @booxter, what do you say of this pr?

I think the above is enough justification.

**Comment by booxter** [comment]
@greg-bock I guess my concern is not exactly about supporting the new feature in kubevirt API, but about the fact that the issue that you pinpointed with having a tangling unused interface inside the launcher pod would still be present when ```authattachPortInterface: false```. While the new Multus feature allows to avoid attaching a pod to the default network, it doesn't allow to have a completely detached launcher pod. (At least one network is always present in the pod; maybe it's CNI API limitation though.)

Anyway, it's indeed not the direct concern for what you try to achieve with this PR.

File: pkg/api/v1/schema.go [hunk 1]
```diff
@@ -920,4 +920,9 @@ type CniNetwork struct {
 	// specified, VMI namespace is assumed.
 	// In case of genie, it references the CNI plugin name.
 	NetworkName string `json:"networkName"`
+
+	// For Multus CNI select the default network and add it to the
+	// multus-cni.io/default-network annotation. Ignored for all
```
Comment (booxter) [inline pkg/api/v1/schema.go]: it's not ignored, it is actually forbidden by the validating hook

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go [hunk 2]
```diff
@@ -1487,6 +1487,77 @@ var _ = Describe("Validating Webhook", func() {
 			causes := ValidateVirtualMachineInstanceSpec(k8sfield.NewPath("fake"), &vm.Spec)
 			Expect(causes).To(BeEmpty())
 		})
+		It("should allow single multus network with a multus default", func() {
+			vm := v1.NewMinimalVMI("testvm")
+			vm.Spec.Domain.Devices.Interfaces = []v1.Interface{
+				*v1.DefaultNetworkInterface(),
+			}
+			vm.Spec.Domain.Devices.Interfaces[0].Name = "multus1"
+			// 3rd interfaces uses the default pod network, name is "default"
```
Comment (booxter) [inline pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go]: I am not sure what this comment here and below tries to communicate

**Comment by greg-bock** [comment]
> While the new Multus feature allows to avoid attaching a pod to the default network, it doesn't allow to have a completely detached launcher pod. (At least one network is always present in the pod; maybe it's CNI API limitation though.)

In our case this is an SRIOV interface.

File: pkg/api/v1/schema.go (hunk 1 above)
Comment (greg-bock) [reply pkg/api/v1/schema.go]: Originally put here before I added that logic. I'll update the text.

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go (hunk 2 above)
Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go]: Another team member wrote these tests. I'll dig around and see if I can figure it out or bug them.

File: pkg/api/v1/schema.go (hunk 1 above)
Comment (greg-bock) [reply pkg/api/v1/schema.go]: Fixed.

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go (hunk 2 above)
Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook_test.go]: I actually might have written these. This was a copy/paste from the test above it that these were based on.

Fixed

**Comment by booxter** [comment]
@greg-bock all good but I think you need to regenerate autogenerated files like swagger api because of the comment changed.

**Comment by greg-bock** [comment]
> Please do make generate and commit changes.

Sorry about that, hadn't yet had a chance to check the CI run from last night. Fixing now.

**Comment by fabiand** [comment]
Gentlereminder to also updat ethe user-guide once this PR is getting merged.

**Comment by greg-bock** [comment]
Sure thing.

**Comment by SchSeba** [comment]
Hi @greg-bock I will like to wait for this PR to be merge https://github.com/kubevirt/kubevirt/pull/2024.

With that PR merge we are going to be able to test this change so can you please add some functional tests to cover this scenario ?

**Comment by greg-bock** [comment]
Sure thing.

**Comment by booxter** [comment]
/hold

@greg-bock in the meantime, could you please prepare the docs patch?

**Comment by greg-bock** [comment]
> @greg-bock in the meantime, could you please prepare the docs patch?

I'll try to get this done tonight and the func test as soon as I can.

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook.go [hunk 3]
```diff
@@ -905,6 +913,22 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 			networkNameMap[spec.Networks[idx].Name] = &spec.Networks[idx]
 		}
 
+		if multusDefaultCount > 1 {
```
Comment (SchSeba) [inline pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: I think you also need to add a validation for pod network.

we should reject a creation of a vm/vmi with pod network and also multus default.

Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: You could have multus as default and still have the pod network.

Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: I think I see what you mean now, it shouldn't named by default.

Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: I actually can't find anything in the code that cares about the pod network's name (I had thought there was some logic there).  Any default logic I could find was based of the interface name being eth0 and that is already handled by the changes.

Comment (SchSeba) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: Hi @greg-bock Is not about the name is about the function I think we should not allow this configuration

```yaml
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  labels:
    special: test
  name: test
spec:
  domain:
    devices:
      interfaces:
      - bridge: {}
        name: default
      - sriov: {}
        name: multus-net
  networks:
  - name: default
    pod: {}
  - multus:
      networkName: sriov
      default: true
    name: multus-net
```

The problem is pod network(eth0) will be connected using the bridge and the multus default (also eth0) will pass the device.

so I think we just need to reject the combination of pod network and default multus network.

Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: Pod network would end up being eth1 in that instance iirc, I'll have to go dig back through the code. We should absolutely allow a Multus network as default and still be able to attach the pod network.

Comment (greg-bock) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: I see where the current implementation would fail now. I'll add this to the validation webhook and make sure I add an example to docs for non pod default with pod as additional CNI network.

File: pkg/api/v1/schema.go [hunk 4]
```diff
@@ -920,4 +920,9 @@ type CniNetwork struct {
 	// specified, VMI namespace is assumed.
 	// In case of genie, it references the CNI plugin name.
 	NetworkName string `json:"networkName"`
+
+	// For Multus CNI select the default network and add it to the
+	// multus-cni.io/default-network annotation. Forbidden for all
+	// other CNIs.
+	Default bool `json:"default,omitempty"`
```
Comment (davidvossel) [inline pkg/api/v1/schema.go]: If this option only pertains to multus, lets just make it only pertain to multus and not create a caveat in the documentation. 

The underlying CniNetwork struct can be broken up like this without impacting backwards compatibility.

```
type MultusNetwork struct {	
	NetworkName string `json:"networkName"`		NetworkName string `json:"networkName"`
        Default bool `json:"default,omitempty"`
}
type GenieNetwork struct {	
	NetworkName string `json:"networkName"`		NetworkName string `json:"networkName"`
}
```

**Comment by davidvossel** [comment]
> If I imagine that we have multiple multus networks, then how many of them need to have set "default: true" in order to provide the default pod network?

The way he has the API now seems pretty close. Only one multus network can have default:true set. That network then becomes the default eth0 for the pod.  We just need validation to ensure we do the right thing when the pod{} network is defined as well.

File: pkg/api/v1/schema.go (hunk 4 above)
Comment (greg-bock) [reply pkg/api/v1/schema.go]: This looked more daunting the first time around when I simply grep'd for all instances of CniNetwork, I now see that this was mostly littered throughout the tests (plus I'm much more familiar with the code than I was a few months ago).

**Comment by greg-bock** [comment]
> * validation webhook that catches if someone tries to set pod{} default when a multus network has default: true set.

I committed this right before your comment. 

> * limit "default" to multus only by not even letting default be in the struct for genie networks.

Updated and commented on the file comment.

**Comment by greg-bock** [comment]
> The way he has the API now seems pretty close. Only one multus network can have default:true set. That network then becomes the default eth0 for the pod. We just need validation to ensure we do the right thing when the pod{} network is defined as well.

I think you can just define the pod network as a multus network with the name of whatever clusterNetwork is set to. I'll try to find some time to test this.

I'll also see if I can test what the behavior is for a normal k8s pod + multus default and make sure it matches. If not it shouldn't be too hard to add some logic around network.Pod to get it to line up. I think the only place eth0 gets set for pod network is [here](https://github.com/kubevirt/kubevirt/blob/1c4f41158c6786ab89f224fb5589d66a6c030162/pkg/virt-launcher/virtwrap/api/converter.go#L1076)

**Comment by SchSeba** [comment]
Hi @greg-bock my pr was merged can you please rebase to use the new multus providers so you will be able to run functional test

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by greg-bock** [comment]
> Hi @greg-bock my pr was merged can you please rebase to use the new multus providers so you will be able to run functional test

Sure thing, it might take me a bit to start work on the functional test but i'll get master merged in now.

**Comment by greg-bock** [comment]
/hold

Holding until I can test non pod default with pod network.

**Comment by greg-bock** [comment]
It appears that using pod with default is not possible at all. This may just be a result of naming conventions for the cluster network or an over zealous regex validation:

```
Warning   FailedCreatePodSandBox   Pod    Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "d15631fea4704c3bc57a04ef2191d1484af46eeb0f18d550b75df39c8b755345" network for pod "pod-case-02": NetworkPlugin cni failed to set up pod "pod-case-02_default" network: Multus: Err in loading K8s Delegates k8s args: Multus: Err in getting k8s network from pod: parsePodNetworkAnnotation: Failed to parse: one or more items did not match comma-delimited format (must consist of lower case alphanumeric characters). Must start and end with an alphanumeric character), mismatch @ 'flannel.1', failed to clean up sandbox container "d15631fea4704c3bc57a04ef2191d1484af46eeb0f18d550b75df39c8b755345" network for pod "pod-case-02": NetworkPlugin cni failed to teardown pod "pod-case-02_default" network: Multus: Err in getting k8s network from pod: parsePodNetworkAnnotation: Failed to parse: one or more items did not match comma-delimited format (must consist of lower case alphanumeric characters). Must start and end with an alphanumeric character), mismatch @ 'flannel.1']
```

It also looks like the version multus is now at in master introduced some changes that break multus looking for the default network in the namespace it is operating in (or this never worked at all). Ref: [254](https://github.com/intel/multus-cni/pull/254/)

I'll test before this multus commit or add logic to check the current ns first if possible and submit an upstream PR.

**Comment by greg-bock** [comment]
I was being dumb and forgot I could pass namespace/network for network name.

**Comment by greg-bock** [comment]
ci test please

**Comment by greg-bock** [comment]
ci test please

**Comment by slintes** [comment]
ci add to whitelist

File: tests/vmi_multus_test.go [hunk 5]
```diff
@@ -407,6 +408,51 @@ var _ = Describe("Multus", func() {
...
+		AfterEach(func() {
+			virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(detachedVMI.Name, &v13.DeleteOptions{})
+			fmt.Printf("Waiting for vmi %s in %s namespace to be removed, this can take a while ...\n", detachedVMI.Name, tests.NamespaceTestDefault)
+			EventuallyWithOffset(1, func() bool {
+				return errors.IsNotFound(virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(detachedVMI.Name, nil))
+			}, 180*time.Second, 1*time.Second).
+				Should(BeTrue())
+		})
+
+		It("should create a virtual machine with one interface with multus default network definition", func() {
```
Comment (SchSeba) [inline tests/vmi_multus_test.go]: @greg-bock nice test!

Comment (greg-bock) [reply tests/vmi_multus_test.go]: I copy/paste well ;)

**Comment by greg-bock** [comment]
> General problem: I think k8s request one IP to return from the cni execution and if we are going to use a L2 or an sriov network for the default one it can be a problem.

The SRIOV case should be fine as it will still request an IP and it is why I opened this PR in the first place. When SRIOV support moves to VFIO only that could change. I started with OVS but was having issues and moved on to PTP. I had quite a time getting a non remote test environment going so I shrugged it off. I'll dig into it as soon as I can.

**Comment by greg-bock** [comment]
@SchSeba you are correct: 

```
Feb 27 17:29:19 node01 kubelet[1620]: W0227 17:29:19.728866    1620 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "virt-launcher-testvmir9dbw9hl4vkk47n4hc4tzqqqwfzzkvxk5kmvcp97rx_kubevirt-test-default": Unexpected address output
Feb 27 17:29:20 node01 kubelet[1620]: E0227 17:29:20.949054    1620 remote_runtime.go:336] ExecSync e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e 'cat /var/run/kubevirt-infra/healthy' from runtime service failed: rpc error: code = Unknown desc = container not running (e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e)
Feb 27 17:29:20 node01 kubelet[1620]: E0227 17:29:20.950685    1620 remote_runtime.go:336] ExecSync e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e 'cat /var/run/kubevirt-infra/healthy' from runtime service failed: rpc error: code = Unknown desc = container not running (e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e)
Feb 27 17:29:20 node01 kubelet[1620]: E0227 17:29:20.951876    1620 remote_runtime.go:336] ExecSync e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e 'cat /var/run/kubevirt-infra/healthy' from runtime service failed: rpc error: code = Unknown desc = container not running (e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e)
Feb 27 17:29:20 node01 kubelet[1620]: W0227 17:29:20.951919    1620 prober.go:103] No ref for container "docker://e9e0c2f170ac410542adb4c77ecc0f258c82ec594e268c6fe0a0e317ac857b5e" (virt-launcher-testvmir9dbw9hl4vkk47n4hc4tzqqqwfzzkvxk5kmvcp97rx_kubevirt-test-default(31c89972-3ab5-11e9-8e10-525500d15501):compute)
```

```    /root/go/src/kubevirt.io/kubevirt/tests/vmi_multus_test.go:472

    VMI unexpectedly stopped. State: Failed
    Expected
        <bool>: true
    to be false
```

Should I add updates the documentation stating that this is not supported. I'm not sure how/if this could be programmatically detected and forbidden in validating-webhook.

**Comment by SchSeba** [comment]
Hi @greg-bock please take a look on this issue https://github.com/kubevirt/kubevirt/issues/2076

I don't think we can create a pod without any interface in the pod network namespace.

**Comment by greg-bock** [comment]
> Hi @greg-bock please take a look on this issue #2076
That is going to be an issue regardless of this new feature from what I read. 

> I don't think we can create a pod without any interface in the pod network namespace.
We've been doing this for months with SRIOV and my last few PRs (this being the last that needs to be upstreamed). I'm think we've tested a kubelet restart to see behavior (cri-o), but it has been a while. We are also looking to possibly switching to containerd but haven't tested that either.

**Comment by SchSeba** [comment]
Hi @greg-bock thanks for the comment.
Can you please replay on the issue if crio support the kubelet restart without eth0 on the pod?

**Comment by greg-bock** [comment]
A VMI with a non bridged default interface, in this case SRIOV, and CRI-O as the runtime is not affected by a kubelet restart.

**Comment by SchSeba** [comment]
Hi @greg-bock  related to your answer just to be sure 

> A VMI with a non bridged default interface, in this case SRIOV, and CRI-O as the runtime is not affected by a kubelet restart.

this mean inside the pod you don't have any interface (only lo)?

**Comment by greg-bock** [comment]
> this mean inside the pod you don't have any interface (only lo)?

Yes

**Comment by SchSeba** [comment]
@thanks @greg-bock !

@rmohr @fabiand @phoracek 
https://github.com/kubevirt/kubevirt/pull/1924#issuecomment-470690457

**Comment by greg-bock** [comment]
ci test please

**Comment by greg-bock** [comment]
None of the failing tests in check-patch.k8s-multus-1.13.3 are related to this change and documentation should good to go. Am I missing anything else?

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 6]
```diff
@@ -1073,7 +1080,11 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 					prefix := ""
 					// no error check, we assume that CNI type was set correctly
 					if net.Multus != nil {
-						prefix = "net"
+						if net.Multus.Default {
```
Comment (booxter) [inline pkg/virt-launcher/virtwrap/api/converter.go]: (I am not exactly sure about it but) Now that we use JSON format to pass networks to CNI, and the JSON dicts contain `name` attributes which list interface names, do we really need to handle `eth` interface prefix here? (This is not a request to change, just a question. If we don't need it, we can clean up in a follow-up.)

Comment (greg-bock) [reply pkg/virt-launcher/virtwrap/api/converter.go]: Can you point to commit/pr where this change was made?

**Comment by booxter** [comment]
@greg-bock sorry for the mess, looks like a simple git conflict resolution was not enough. It fails to build a test.

```
/root/.cache/bazel/_bazel_root/6f347497f91c9a385dcd9294645b76e0/sandbox/processwrapper-sandbox/867/execroot/__main__/pkg/virt-controller/services/template_test.go:176:19: undefined: "kubevirt.io/kubevirt/pkg/api/v1".CniNetwork
```

**Comment by greg-bock** [comment]
I'll force push back to my last commit and look at it tonight.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 6 above)
Comment (greg-bock) [reply pkg/virt-launcher/virtwrap/api/converter.go]: I see it now.

Comment (booxter) [reply pkg/virt-launcher/virtwrap/api/converter.go]: @greg-bock here https://github.com/kubevirt/kubevirt/pull/2107 it was merged a day ago.

**Comment by greg-bock** [comment]
Looks like I need to fix my output checking tests @booxter. Would it be worthwhile to try to drop the naming logic before merging? It might take me a few days to see/test if I can use what is present in the JSON vs the loop/count logic currently in use.

File: pkg/virt-controller/services/template_test.go [hunk 7]
```diff
@@ -190,6 +190,37 @@ var _ = Describe("Template", func() {
...
+				}
+
+				pod, err := svc.RenderLaunchManifest(&vmi)
+				Expect(err).ToNot(HaveOccurred())
+				value, ok := pod.Annotations["v1.multus-cni.io/default-network"]
+				Expect(ok).To(Equal(true))
+				Expect(value).To(Equal("default"))
+				value, ok = pod.Annotations["k8s.v1.cni.cncf.io/networks"]
+				Expect(ok).To(Equal(true))
+				Expect(value).To(Equal("[{\"interface\":\"net1\",\"name\":\"test1\",\"namespace\":\"default\"}]"))
```
Comment (booxter) [inline pkg/virt-controller/services/template_test.go]: hm, shouldn't both networks be defined here? For example, if user would like to specify a custom MAC address for the default Multus network, shouldn't they have the MAC address passed through this JSON dict?

Comment (SchSeba) [reply pkg/virt-controller/services/template_test.go]: @booxter This is not supported yet for default networks I create a PR in multus https://github.com/intel/multus-cni/pull/268 to allow default network to have a json struct an not only a name.

@booxter , @greg-bock please review it and comment there if you find that use full to get it merge.

**Comment by booxter** [comment]
@SchSeba is there anything else missing here, or we can finally merge it? The user-guide update seems fine too.

If I don't hear back before then I plan to merge this tomorrow. ;)

**Comment by booxter** [comment]
I realize folks from Israel are not online before Sunday so postponing merge till Monday to give them a chance to chime in.
